Mathlib/Algebra/Module/Length.lean

```diff
@@ -0,0 +1,931 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.Algebra.Module.Submodule.JordanHolder
+import Mathlib.RingTheory.Noetherian
+import Mathlib.RingTheory.Artinian
+import Mathlib.Order.KrullDimension
+import Mathlib.Algebra.BigOperators.Order
+import Mathlib.LinearAlgebra.Basis.VectorSpace
+import Mathlib.LinearAlgebra.FreeModule.Finite.Basic
+import Mathlib.Algebra.BigOperators.WithTop
+
+/-!
+
+# Module Length
+
+for an `R`-module `M`, if it has a composition series `0 ⋖ ... ⋖ M`, then its length is defined by
+the length of this composition series, otherwise its length is said to be infinite. By Jordan-Hölder
+theorem this is well defined.
+
+Some authors also define the length of `M` to be the length of longest `N₀ < N₁ < ... < Nₖ`. In this
+file, we show that two definitions are in fact equal.
+
+-/
+
+variable {R : Type*} [CommRing R] {M : Type*} [AddCommGroup M] [Module R M]
+
+open Classical BigOperators
+
+variable (R M)
+
+/-- A module with finite length is a module with a composition series starting with 0 and ending
+  with itself. -/
+class FiniteLengthModule where
+  /-- A finite length module admits a composition series `s`. -/
+  compositionSeries : CompositionSeries (Submodule R M)
+  /-- the leftest element is the 0 submodule -/
+  head_eq : compositionSeries.head = ⊥
+  /-- the rightest element is the whole thing -/
+  last_eq : compositionSeries.last = ⊤
```

Comment:
I feel that we should define FiniteLength for any JordanHolderLattice, and we should be able to show a lattice is finite length iff both < and > are well-founded. Also I think most of the places there's no reason to assume R to be commutative.

Reply:
Good point! I will try to generalize finite length to all jordan holder lattices! Thanks for the suggestions.

It is probably also true that we can get pass by assume M is `AddCommMonoid` and R `semiring` even.